metricbeat/module/elasticsearch/index/data.go

```diff
@@ -19,70 +19,379 @@ package index
+	TotalSizeInBytes  int `json:"total_size_in_bytes"`
+	AvgTimeInMillis   int `json:"avg_time_in_millis"`
+	AvgSizeInBytes    int `json:"avg_size_in_bytes"`
+}
+
+func eventsMapping(r mb.ReporterV2, m *MetricSet, info elasticsearch.Info, content []byte) error {
+	clusterStateMetrics := []string{"metadata", "routing_table"}
+	clusterState, err := elasticsearch.GetClusterState(m.HTTP, m.HTTP.GetURI(), clusterStateMetrics)
+	if err != nil {
+		return errors.Wrap(err, "failure retrieving cluster state from Elasticsearch")
 	}
-)
 
-func eventsMapping(r mb.ReporterV2, info elasticsearch.Info, content []byte) error {
-	var indicesStruct IndicesStruct
-	err := json.Unmarshal(content, &indicesStruct)
+	var indicesStats stats
+	if err := parseAPIResponse(content, &indicesStats); err != nil {
+		return errors.Wrap(err, "failure parsing Indices Stats Elasticsearch API response")
+	}
+
+	indicesSettings, err := elasticsearch.GetIndicesSettings(m.HTTP, m.HTTP.GetURI())
 	if err != nil {
-		return errors.Wrap(err, "failure parsing Elasticsearch Stats API response")
+		return errors.Wrap(err, "failure retrieving indices settings from Elasticsearch")
 	}
 
 	var errs multierror.Errors
-	for name, index := range indicesStruct.Indices {
-		event := mb.Event{}
+	for name, idx := range indicesStats.Indices {
+		event := mb.Event{
+			ModuleFields: common.MapStr{},
+		}
+		idx.Index = name
 
-		event.RootFields = common.MapStr{}
-		event.RootFields.Put("service.name", elasticsearch.ModuleName)
+		settings, exists := indicesSettings[name]
+		if exists {
+			idx.Hidden = settings.Hidden
+		}
+
+		err = addClusterStateFields(&idx, clusterState)
+		if err != nil {
+			errs = append(errs, errors.Wrap(err, "failure adding cluster state fields"))
+			continue
+		}
 
-		event.ModuleFields = common.MapStr{}
-		event.ModuleFields.Put("cluster.name", info.ClusterName)
 		event.ModuleFields.Put("cluster.id", info.ClusterID)
+		event.ModuleFields.Put("cluster.name", info.ClusterName)
 
-		event.MetricSetFields, err = schema.Apply(index)
+		// Convert struct to common.Mapstr by passing it to JSON first so we can store the data in the root of the
+		// metricset level
+		byt2, err := json.Marshal(idx)
 		if err != nil {
-			errs = append(errs, errors.Wrap(err, "failure applying index schema"))
+			errs = append(errs, errors.Wrap(err, "failure trying to convert metrics results to JSON"))
 			continue
 		}
-		// Write name here as full name only available as key
-		event.MetricSetFields["name"] = name
+		var indexOutput common.MapStr
+		if err = json.Unmarshal(byt2, &indexOutput); err != nil {
+			errs = append(errs, errors.Wrap(err, "failure trying to convert JSON metrics back to mapstr"))
+			continue
+		}
+
+
+		event.MetricSetFields = indexOutput
+		event.MetricSetFields.Put("name", name)
+		event.MetricSetFields.Put("total.store.size.bytes", idx.Total.Store.SizeInBytes)
+		event.MetricSetFields.Put("total.segments.memory.bytes", idx.Total.Segments.MemoryInBytes)
```

Comment:
I'm curious why these two fields need special treatment here. Shouldn't they get marshalled & unmarshalled above along with all the other fields?

Reply:
Frankly, I don't remember either but I have double checked it and I think that it was an attempt to rename the fields to follow the naming conventions, that I later realized that I had way too many to change and then I regret and I forgot to remove those two.

Reply:
I finally removed them

Reply:
Ha! No, forget that. They were original fields https://github.com/elastic/beats/pull/21438/files#diff-397cb07191b7b507fd7ca773b66d961954da4c2bfc5694ab8cf997e1cf991037R44

Instead of creating an alias I just duplicated them. I'll add the alias.